### PR TITLE
feat: Add functions and prebuiltUI for oauth2 logout

### PR DESCRIPTION
**Pull request by anku255**
## Summary of change

This PR adds functions and prebuiltUI for oauth2 logout.

## Related issues

-   https://github.com/supertokens/supertokens-auth-react/issues/847

## Test Plan

(Write your test plan here. If you changed any code, please provide us with clear instructions on how you verified your changes work. Bonus points for screenshots and videos!)

## Documentation changes

(If relevant, please create a PR in our [docs repo](https://github.com/supertokens/docs), or create a checklist here highlighting the necessary changes)

## Checklist for important updates

-   [ ] Changelog has been updated
-   [ ] `frontendDriverInterfaceSupported.json` file has been updated (if needed)
-   [ ] Changes to the version if needed
    -   In `package.json`
    -   In `package-lock.json`
    -   In `lib/ts/version.ts`
-   [ ] Had run `npm run build-pretty`
-   [ ] Had installed and ran the pre-commit hook
-   [ ] Issue this PR against the latest non released version branch.
    -   To know which one it is, run find the latest released tag (`git tag`) in the format `vX.Y.Z`, and then find the latest branch (`git branch --all`) whose `X.Y` is greater than the latest released tag.
    -   If no such branch exists, then create one from the latest released branch.
-   [ ] If added a new recipe interface, then make sure that the implementation of it uses NON arrow functions only (like `someFunc: function () {..}`).
-   [ ] If I added a new recipe, I also added the recipe entry point into the `size-limit` section of `package.json` with the size limit set to the current size rounded up.
-   [ ] If I added a new recipe, I also added the recipe entry point into the `rollup.config.mjs`
-   [ ] If I added a new login method, I modified the list in `lib/ts/types.ts`
-   [ ] If I added a factor id, I modified the list in `lib/ts/recipe/multifactorauth/types.ts`

### CODE REVIEW

File: lib/ts/recipe/oauth2provider/types.ts [hunk 1]
```diff
@@ -45,17 +47,39 @@ export type ContinueOAuth2AfterRefreshRedirectContext = {
...
+          userContext: UserContext;
+      }
+    | {
+          action: "OAUTH2_LOGOUT_SUCCESS";
+          frontendRedirectTo: string;
+          userContext: UserContext;
+      };
+
 // TODO: update this whenever we add components
 export type ComponentOverrideMap = any;
```
[the remark this answers is not in the document]
Comment (anku255) [reply lib/ts/recipe/oauth2provider/types.ts]: Done

File: lib/ts/recipe/oauth2provider/components/features/oauth2LogoutScreen/index.tsx [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+
+    const logoutChallenge = getQueryParams("logoutChallenge") ?? undefined;
+    const navigate = props.navigate ?? UI.getReactRouterDomWithCustomHistory()?.useHistoryCustom();
+
+    const onLogout = React.useCallback(async () => {
+        if (logoutChallenge === undefined) {
+            return;
+        }
+        setIsLoggingOut(true);
+        const { frontendRedirectTo } = await props.recipe.logOut(logoutChallenge, userContext);
```
[the remark this answers is not in the document]
Comment (anku255) [reply lib/ts/recipe/oauth2provider/components/features/oauth2LogoutScreen/index.tsx]: Done

File: lib/ts/recipe/oauth2provider/components/themes/oauth2LogoutScreen/index.tsx [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+                        onClick={props.onLogoutClicked}
+                    />
+                    <SuperTokensBranding />
+                </div>
+            </div>
+        </div>
+    );
+};
+
+const OAuth2LogoutThemeWithOverride = withOverride("OAuth2LogoutScreen", OAuth2LogoutScreen);
```
[the remark this answers is not in the document]
Comment (anku255) [reply lib/ts/recipe/oauth2provider/components/themes/oauth2LogoutScreen/index.tsx]: Done

File: lib/ts/recipe/oauth2provider/components/themes/oauth2LogoutScreen/index.tsx [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+const OAuth2LogoutScreen: FC<OAuth2LogoutScreenThemeProps> = (props) => {
+    const t = useTranslation();
+
+    if (props.showSpinner) {
+        return <DynamicLoginMethodsSpinner />;
+    }
+
+    return (
+        <div data-supertokens="oauth2Logout">
+            <div data-supertokens="container">
```
[the remark this answers is not in the document]
Comment (anku255) [reply lib/ts/recipe/oauth2provider/components/themes/oauth2LogoutScreen/index.tsx]: done

File: lib/ts/recipe/oauth2provider/prebuiltui.tsx [hunk 5]
```diff
@@ -64,25 +65,52 @@ export class OAuth2ProviderPreBuiltUI extends RecipeRouter {
                 recipeID: OAuth2ProviderRecipe.RECIPE_ID,
             };
         }
+        // TODO: Should this be checking this.recipeInstance.config.oauth2LogoutScreen.disableDefaultUI
```
[the remark this answers is not in the document]
Comment (anku255) [reply lib/ts/recipe/oauth2provider/prebuiltui.tsx]: done

File: lib/ts/recipe/oauth2provider/recipe.ts [hunk 6]
```diff
@@ -116,11 +121,24 @@ export default class OAuth2Provider extends RecipeModule<
...
+        userContext: UserContext
+    ): Promise<{
+        status: "OK";
+        frontendRedirectTo: string;
+        fetchResponse: Response;
+    }> {
+        return await this.webJSRecipe.logOut({ logoutChallenge, userContext });
+    }
+
     /*
```
[the remark this answers is not in the document]
Comment (anku255) [reply lib/ts/recipe/oauth2provider/recipe.ts]: Fixed

File: test/end-to-end/oauth2provider.test.js [hunk 7]
```diff
@@ -167,6 +168,67 @@ describe("SuperTokens OAuth2Provider", function () {
...
+            await stLogoutButton.click();
+
+            // Ensure the final url matches the post_logout_redirect uri
+            await waitForUrl(page, "/oauth/login?logout=true", false);
+
+            await page.waitForSelector("#oauth2-token-data", { hidden: true });
+
+            // Ensure supertokens session was cleared
+            const sessionInfoRes = await fetch(`${TEST_APPLICATION_SERVER_BASE_URL}/sessioninfo`);
+            assert.strictEqual(sessionInfoRes.status, 401);
```
Comment (porcellus) [inline test/end-to-end/oauth2provider.test.js:229]: AFAIK, this wouldn't fail even if the session wasn't cleared in the browser, as this fetch is running in the test runner env (node).
This needs to be checked on the frontend (e.g.: you can start the login flow again and see that it redirects to the auth screen)

**Review by porcellus — CHANGES_REQUESTED** [review]

File: test/end-to-end/oauth2provider.test.js (hunk 7 above)
Comment (anku255) [reply test/end-to-end/oauth2provider.test.js]: My bad!  I have updated it to check for `/auth` redirection after clicking on login. We could have also did the fetch call in `page.evaluate`, I guess.

**Review by porcellus — APPROVED** [review]

**Pull request merged**